File: spark-all-demo/spark-scala-demo/src/main/scala/com/zgb/collection/MyFunction2.java
package com.zgb.collection;

/**
 * @author xmly
 * @email guanbao.zhou@ximalaya.com
 * @Date 2021/3/19 9:00 下午
 * @Created By guanbao.zhou
 */
public interface MyFunction2<T, W, R> {
  R apply(T t, W w);
}

File: spark-all-demo/spark-scala-demo/src/main/scala/com/zgb/kl/MyFunction2.scala
package com.zgb.kl

/**
 * @email guanbao.zhou@ximalaya.com
 * @Date 2021/3/20 10:50 下午
 * @Created By guanbao.zhou
 * @author xmly
 * Scala的泛型，scala的泛型用[]
 * 逆变：-A 方法的输入参数类型
 * 协变：+R 方法返回值的类型
 */
trait MyFunction2[-A, -B , +C] {
  def apply(a: A, b: B) : C
}
